qcog_python_client/qcog/pytorch/agent.py

```diff
@@ -0,0 +1,140 @@
+        tool = self.tools.get(tool_name)
+
+        if not tool:
+            raise ValueError(
+                f"Tool {tool_name} not found. Available tools: {list(self.tools.keys())}"  # noqa
+            )
+
+        return tool
+
+    def set_context(self, key: str, value: Any) -> None:
+        """Set a value on the chain context."""
+        self.chain.context[key] = value
+
+    def init(self, *chain: Handler) -> Handler:
+        """Initialize the Responsibility Chain."""
+        if len(chain) == 1:
+            raise ValueError("Chain must have at least 2 handlers.")
+        head = chain[0]
+
+        context: dict = {}
+
+        for i in range(len(chain) - 1):
+            current_handler = chain[i]
+            next_handler = chain[i + 1]
+            current_handler.set_next(next_handler, head)
+
+        # Register the get_tool function
+        for handler in chain:
+            handler.get_tool = self._get_tool
+            handler.context = context
+
+        self._chain = head
+        return head
+
+    async def upload_model(self, model_path: str, model_name: str) -> dict:
+        """Upload the model to the server."""
+        # Init Command will dispatch a Discover Command
+
+        handler = await self.chain.dispatch(
+            payload=DiscoverCommand(
+                model_name=model_name,
+                model_path=model_path,
+                dispatch_next=True,  # Will follow the whole chain
+            )
+        )
+        upload_handler = cast(UploadHandler, handler)
+
+        # We dont wanna expose the chain or the handlers outside of the PytorchAgent
+        return upload_handler.created_model
+
+    async def train_model(
+        self,
+        model_guid: str,
+        *,
+        dataset_guid: str,
+        training_parameters: dict,
+    ) -> dict:
+        """Train the model."""
+        # Upload training command
+
+        handler = await self.chain.dispatch(
+            payload=SaveParametersCommand(parameters=training_parameters)
+        )
+        save_parameters_handler = cast(SaveParametersHandler, handler)
+
+        training_parameters_guid = save_parameters_handler.parameters_response["guid"]
+
+        train_handler = await self.chain.dispatch(
+            payload=TrainCommand(
+                model_guid=model_guid,
+                dataset_guid=dataset_guid,
+                training_parameters_guid=training_parameters_guid,
+            )
+        )
+
+        handler = cast(TrainHandler, train_handler)
+        return handler.trained_model
+
+    async def inference(self, data: Any, model_name: str) -> Any:
+        """Run inference."""
+        raise NotImplementedError()
+
+    def register_tool(
+        self,
+        tool_name: str,
+        fn: ToolFn,
+    ) -> None:
+        """Register a tool to be used by the handlers in the chain."""
+        self.tools[tool_name] = fn
+
+    @classmethod
+    def create_agent(
+        cls,
+    ) -> "PyTorchAgent":
```

Comment:
`from __future__ import annotations` at the top and you'll be able to directly reference `PyTorchAgent` in your typehint